src/filter.rs

```diff
@@ -105,152 +105,90 @@ pub fn separable_filter<I, K>(image: &I, h_kernel: &[K], v_kernel: &[K])
+    assert!(kernel.len() as u32 >= k_width * k_height,
+            "incompatible kernel lengths".to_owned());
 
-    // TODO: Should we handle images with height or width < 2? Feels clunky to return Results
-    // TODO: everywhere. Could just document as requirement and leave it at that.
-    for y in 0..height {
-        let y_prev = cmp::max(1, y) - 1;
-        let y_next = cmp::min(height - 2, y) + 1;
-
-        for x in 0..width {
-            let x_prev = cmp::max(1, x) - 1;
-            let x_next = cmp::min(width - 2, x) + 1;
-
-            let mut acc = vec![Zero::zero(); num_channels];
-            let pixels = unsafe {
-                [&image.unsafe_get_pixel(x_prev, y_prev),
-                 &image.unsafe_get_pixel(x     , y_prev),
-                 &image.unsafe_get_pixel(x_next, y_prev),
-                 &image.unsafe_get_pixel(x_prev, y     ),
-                 &image.unsafe_get_pixel(x     , y     ),
-                 &image.unsafe_get_pixel(x_next, y     ),
-                 &image.unsafe_get_pixel(x_prev, y_next),
-                 &image.unsafe_get_pixel(x     , y_next),
-                 &image.unsafe_get_pixel(x_next, y_next),]
-            };
-            for i in 0..9 {
-                accumulate(&mut acc, pixels[i], kernel[i]);
+    let (w, h) = image.dimensions();
+    let mut out = VecBuffer::<Q>::new(w, h);
+    let num_channels = I::Pixel::channel_count() as usize;
+    let zero = K::zero();
+    let mut acc = vec![zero; num_channels];
+
+    for y in 0..h {
+        for x in 0..w {
+            for k_y in 0..k_height {
+                let y_p = cmp::min(h + h - 1, cmp::max(h, (h + y + k_y - k_height / 2))) - h;
+                for k_x in 0..k_width {
+                    let x_p = cmp::min(w + w - 1, cmp::max(w, (w + x + k_x - k_width / 2))) - w;
+                    let (p, k) = unsafe {
+                        (image.unsafe_get_pixel(x_p, y_p),
+                         *kernel.get_unchecked((k_y * k_width + k_x) as usize))
+                    };
+                    accumulate(&mut acc, &p, k);
+                }
+            }
+            let out_channels = out.get_pixel_mut(x, y).channels_mut();
+            for (a, c) in acc.iter_mut().zip(out_channels.iter_mut()) {
+                f(c, a.clone());
+                *a = zero;
             }
-
-            clamp_acc(&acc, out.get_pixel_mut(x, y).channels_mut());
         }
     }
 
     out
 }
 
+/// Returns 2d correlation of an image with a 3x3 kernel. Intermediate calculations are
+/// performed at type K, and the results clamped to subpixel type S. Pads by continuity.
+pub fn filter3x3<I, P, K, S>(image: &I, kernel: &[K]) -> VecBuffer<ChannelMap<P, S>>
+    where I: GenericImage<Pixel=P>,
+          P::Subpixel: ValueInto<K>,
+          S: Clamp<K> + Primitive + 'static,
+          P: WithChannel<S> + 'static,
+          K: Num + Copy {
+    filter_kernel(image, kernel, 3, 3, |channel, acc| *channel = S::clamp(acc))
+}
+
```

Comment:
Sadly, I'm not aware of any existing benchmarks. My plans didn't extend beyond just installing/building some other libraries and writing the same benchmarks we have in this one. I've not yet done this.